File: 프로그래머스/lv2/42842. 카펫/카펫.java
import java.util.ArrayList;

class Solution {
    public int[] solution(int brown, int yellow) {
        int[] answer = new int[2];
        ArrayList<Integer> a = new ArrayList<>();
        ArrayList<Integer> b = new ArrayList<>();
        int total = brown + yellow;
        for (int i =1; i<=Math.sqrt(total); i++){
            if (total%i == 0){
                a.add(i);
                b.add(total/i);
            }
        }

        for (int j = a.size()-1; j>=0; j--){
            if (a.get(j)*b.get(j) == total && (a.get(j)-2)*(b.get(j)-2) == yellow){
                answer[0]=b.get(j);
                answer[1]=a.get(j);
                break;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/42842. 카펫/카펫.py
def solution(brown, yellow):
    total = brown + yellow  
    for x in range(total, 2, -1):  # 가로 길이 x를 전체 격자 수부터 시작하여 감소시키며 탐색
        if total % x == 0:  # x가 전체 격자 수의 약수일 경우
            y = total // x  # 세로 길이 y를 계산
            if (x-2) * (y-2) == yellow:  # 노란색 격자의 수가 조건을 만족하는지 확인
                return [x, y]  # 조건을 만족하는 가로, 세로 길이 반환